### ISSUE TITLE
KafkaConsumerProperties and KafkaProducerProperties configuration randomness

### ISSUE THREAD

**Issue by mehdizyane** [spring-cloud/spring-cloud-stream-binder-kafka#805]
Hello,

In the attachement you'll find anexample of a configuration file who produce some random results, I'm having three kafka.streams.bindings with different custom configuration values in order to configure my custom Kafka Serde.

I've noticed that the "input" configuration is the only one taken into account the others are simply ignored and replaced by the "input" configuration.

In my understanding we can give each binding a proper set of properties, maybe I'm wrong.

Thank you in advance for all your efforts.

[kafka_binder.txt](https://github.com/spring-cloud/spring-cloud-stream-binder-kafka/files/3870412/kafka_binder.txt)

**Comment by sobychacko** [comment]
@mehdizyane Which version of the binder are you using? Can you try it on the `3.0` version which will be GA released next week? If you are still seeing issues, can you provide a small sample application?

**Comment by mehdizyane** [comment]
@sobychacko The version is 2.2.1.RELEASE.

I will try with the 3.0 and I'll get back to you with a sample application.

Thanks.

**Comment by mehdizyane** [comment]
Hello,

In the attachement you'll find a sample application that reproduces the issue with 3.0.0.RC2 version.

In this example I start 2 producers and 1 consumer, the three CustomSerde initiated receives the consumer configuration, the custom producer configuration is just ignored. I've also noticed that all the producers configurations are ignored in all cases.

You'll need a broker started locally.

[spring-cloud-stream-properties-issue.zip](https://github.com/spring-cloud/spring-cloud-stream-binder-kafka/files/3875878/spring-cloud-stream-properties-issue.zip)

**Comment by sobychacko** [comment]
@mehdizyane Thanks for the explanation and the example. When you use `producer.configuration`, the way you are using, it may not work as you expect. Each Kafka Streams processor is attached to a single `StreamsBuilder` by the binder and this `StreamsBuilder` can only be mapped to a single configuration at the individual binding level.
The binder offers to pass `configuration` at the binder level (`spring.cloud.stream.kakfak.streams.binder.configuration`) or binding level (as you have in the example) but if there are multiple bindings in a single processor, usually, the configuration provided at the first binding takes precedence, usually the first consumer binding. This is the reason, why you are seeing that the consumer configuration always takes effect. 

Are you trying to provide different `Serde` implementations for the various producers? 
It is not clear if that is what you are trying to accomplish as I don't see any `Serde` values
in those configurations. Nonetheless, if that is your use case, then you can set those at the producer binding without using the `configuration`. 

For e.g.
```
spring.cloud.stream.kafka.streams.bindings.output1.producer.keySerde: org.apache.kafka.common.serialization.Serdes$StringSerde
spring.cloud.stream.kafka.streams.bindings.output1.producer.valueSerde: com.randomgroupid.serde.SpecialSerde1
spring.cloud.stream.kafka.streams.bindings.output12.producer.keySerde: org.apache.kafka.common.serialization.Serdes$StringSerde
spring.cloud.stream.kafka.streams.bindings.output2.producer.valueSerde: com.randomgroupid.serde.SpecialSerde2
```

The bottom line here is to use the binder level `configuration` or `binding` configuration for **only** providing common elements that will be applied to the `StreamsBuilder`.  For e.g. if you have a single `StreamListener`, use the binder configuration approach, but multiple `StreamListener` methods, use the first input binding to pass the configuration.  

Hope this helps.

**Comment by mehdizyane** [comment]
@sobychacko Thanks for your explanation.

Actually I was trying to put different meta-data to different events comming out from different producers, something like :
- Branch isValid : eventName = validContract
- Branch notValid : eventName = rejectedContract

I was hoping to do so without writing it and just relying on the producer configuration part, the Serde gets the configuration map and passes it to the Serialiazer who puts the metadata on my event before serialization.

Thanks again.
